augur/curate/format_dates.py

```diff
@@ -0,0 +1,237 @@
+#!/usr/bin/env python3
+"""
+Format date fields to ISO 8601 dates (YYYY-MM-DD), where incomplete dates
+are masked with 'XX' (e.g. 2023 -> 2023-XX-XX).
+"""
+import re
+from datetime import datetime
+
+from augur.errors import AugurError
+from augur.io.print import print_err
+from augur.types import DataErrorMethod
+
+
+def register_parser(parent_subparsers):
+    parser = parent_subparsers.add_parser("format-dates",
+        parents=[parent_subparsers.shared_parser],
+        help=__doc__)
+
+    required = parser.add_argument_group(title="REQUIRED")
+    required.add_argument("--date-fields", nargs="+",
+        help="List of date field names in the record that need to be standardized.")
+    required.add_argument("--expected-date-formats", nargs="+",
+        help="Expected date formats that are currently in the provided date fields, " +
+             "defined by standard format codes as listed at " +
+             "https://docs.python.org/3/library/datetime.html#strftime-and-strptime-format-codes. " +
```

Comment:
The standard format codes can't handle partial ambiguity in years (e.g. `201X`, `19XX`). These should be supported for usage such as in `augur refine --year-bounds` (more info in #989).

Reply:
I think the idea here is that `augur curate format-dates` is to convert from a variety of date strings used in the wild into a single standard ISO 8601 form, and that data sources you'd be curating with it are unlikely to be using the Nextstrain-style XX ambiguity.

I think if you needed to support some XX-ambiguity (maybe more limited than what your date parsing improvements support), you could probably pass format strings like `%Y-%m-XX`?

Reply:
Will revisit partial ambiguity as a future improvement (https://github.com/nextstrain/augur/issues/1255)